### DISCUSSION TITLE
JavaScript Macro's (Heavy Calculations) "Return Value" Best Practices

### DISCUSSION THREAD

**Discussion by msameen99** [TiddlyWiki/TiddlyWiki5#6378] [Q&A] [answered]
Hello,

I've a "JavaScript Macro" that do a heavy calculation (inside and outside TW5) and the final calculation result is at least 3 level "deep" JSON object.

For Ex., I'll pass a single parameter  (TopicID) and the return Object in the form of ( --hasMulti--> Books, Course, Tutorials, ... --eachHasMulti--> Authors, Websites, YouTube Channels, .... --eachHasMulti--> Attributes).
Also, I need to work with individual levels (Books, Courses, Authors, Attributes, .....) in different places inside the same Tiddler.

Now, my question is: I'm thinking to return this information to WikiText as a single value string variable (after converting the JSON to a string), is this a good idea? or there are other best practices?
In other words, what is the Best Practices to pass a "Multi Level Related" information to WikiText in a way that work naturally/easy with FILTERS Operators?

Or, Do I need to break that Macro to a several Macros? each will "repeat" the "exact" same "heavy" calculations BUT return a simple "List"?
(as I understand, the "lists" variables and "single value" variables are the 1st citizen in WikiText, all filter operators are working wonderful with them.)

Thanks

**Comment by pmario** [comment]
At the moment there is no easy way to access multi-level JSON objects in wikitext. If you need to work with filters afterwards you have to reduce your structure to a "single level" JSON which basically is a `key: value` relation. 

The wikitext term is "data tiddler".

If it's possible to use data-tiddlers it would be the easiest way to work with filters, since you can work with indexes. `index = key`. Working with data-tiddlers and filters still has some limitations. .. 

So the best way would be to create a structure, that uses tiddler-fields, where the field names are the `key` ... 

If you could be a bit more specific about "heavy calculations" it would be easier to see, what's probably needed, of if there is a structure that would more naturally fit to the TW filter syntax.

**Reply by msameen99** [reply]
Thanks a lot pmario for your reply, 

1- What I know is (please correct me if wrong) the TW5 JavaScript Macros should NOT alter any "states" , so I should not create a new Tiddlers from within Macros. is that correct?

2- Is there any "restrictions" on the size of the returned string from JavsScript Macros? 
This question because I'm thinking to "flatten" all JSON levels to a single string, using the characters "|","||", "|||", ... as a separator between the levels (the return string will not contains that character, it has only some IDs and Tiddlers Titles) BUT this may result a VERY long string.

**Comment by pmario** [comment]
>1- What I know is (please correct me if wrong) the TW5 JavaScript Macros should NOT alter any "states" , so I should not create a new Tiddlers from within Macros. is that correct?

Yes, that's correct. ... Macros don't export a `.refresh()` function, so they need to be part of a widget body, to be refreshed. .. That's the main reason, why macros should not have "side effects". ..

**Comment by pmario** [comment, marked as answer]
>2- Is there any "restrictions" on the size of the returned string from JavsScript Macros?

no

>This question because I'm thinking to "flatten" all JSON levels to a single string, using the characters "|","||", "|||", ... as a separator between the levels (the return string will not contains that character, it has only some IDs and Tiddlers Titles) BUT this may result a VERY long string.

There is a JSON-mangler plugin, which pretty much does what you want. .. BUT it's broken since 5.2.x ...

**Comment by pmario** [comment]
It's actually not that hard to create widgets. Have a look at https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/action-createtiddler.js which is probably very close to what you need to write into a new tiddler. 

action-widgets usually need to be triggered by the `actions` attribute from buttons.

**Reply by msameen99** [reply]
Thanks again pmario, I'll check the "widget" option as well

**Comment by Jermolene** [comment]
Hi @msameen99 one approach that we use in the core is to use a custom widget message to trigger the JS function, and have it return the data as a structure of interlinked tiddlers. Very roughly, you'd have a startup module with code like this:

```js
exports.startup = function() {
	// Listen for widget messages to control loading the plugin library
	$tw.rootWidget.addEventListener("tm-my-message",function(event) {
		var paramObject = event.paramObject || {};
		var tiddlers = myMagicFunction(paramObject.firstParameter,paramObject.secondParameter);
		$tw.wiki.addTiddlers(tiddlers);
	}
};
```

Depending on the precise scenario, it may make sense to have one of the parameters specifying the root tiddler tiddler title for the results.
